tests for chrono datetime parsing (#28)
FlatFilers/platform-sdk-starter: examples/fields/SmartDateField.README.md

# SmartDateField

## Why

> The `SmartDateField` lets you worry about operations on `Date`s as objects, without worrying about parsing dates, or formatting dates.  Currently the `SmartDateField` can parse 9 date types without any format strings.  This lets you concentrate on business logic, without worrying about incoming date formats.  This let's you confidently perform arithmetic and comparisons in `compute`, `recordCompute`, `batchRecordsCompute`, and `validate` on `Date`s.  `SmartDateField` has an `fString` argument that controls how the date is serialized after `validate` is finished.  Flatfile users no longer have to convert in and out of dates to maintain formatting.

Reviewer: What are the 9 date types that can be parsed without any format strings?

Author: They are visible in tests... I'll add a link

## What

There are comprehensive tests for `SmartDateField`, the most demonstrative being [DateSheet.spec.ts](./DateSheet.spec.ts) .  This shows how dates can be compared as first class objects in a `recordCompute` function, and then formatted to a different persistence formate.

### Assumptions made

We will add options to override these assumption as needed.

Date's without an Time of Day (TOD) and without a timezone (TZ)  will be interpreted as 00:00:00 in GMT.  By default JS dates are formatted to 00:00:00 in the local timezone, this leads to inconsistent results based on how the server is setup.  00:00:00 is also the the behavior of pandas timestamp

Dates with a TOD and without a TZ will be interpreted as that TOD in GMT.  You might want to default to a local timezone here, that option will be added.

Dates with a TOD and TZ will be interpreted with that TOD and TZ.

Dates that can't be explicitly parsed will throw an error and remain as the original string.

Please report bugs where this behavior isn't the case.

> ## Installing SmartDateField in an existing Faltfile project

Reviewer: Faltfile -> Flatfile

`SmartDateField` is currently in public beta.  This exact code will work currently and against future versions of Flatfile... but we might update the behavior of the `SmartDateField` that is distributed going forward.  This lets us get feedback from users quickly, without locking us into supporting temporary design decisions.

> First copy the whole of [SmartDateField.ts](./SmartDateField.ts) into your repo.

Reviewer: Where should this be copied to in the repo?

Author: I updated the instructions

# SmartDateField

## Why

The `SmartDateField` lets you worry about operations on `Date`s as objects, without worrying about parsing dates, or formatting dates.  Currently the `SmartDateField` can parse 9 date types without any format strings.  This lets you concentrate on business logic, without worrying about incoming date formats.  This let's you confidently perform arithmetic and comparisons in `compute`, `recordCompute`, `batchRecordsCompute`, and `validate` on `Date`s.  `SmartDateField` has an `fString` argument that controls how the date is serialized after `validate` is finished.  Flatfile users no longer have to convert in and out of dates to maintain formatting.

## What

> There are comprehensive tests for `SmartDateField`, the most demonstrative being [DateSheet.spec.ts](./DateSheet.spec.ts) .  This shows how dates can be compared as first class objects in a `recordCompute` function, and then formatted to a different persistence formate.

Reviewer: Missed the importance of this on the first pass!  Found the `fString` portion of the code pulled the concepts together.  Should that be mentioned here (see it is mentioned in the "Why" section or will someone picking this up understand how they can use `fString`?

Author: Great catch, changed the wording

### Assumptions made

We will add options to override these assumption as needed.

> Date's without an Time of Day (TOD) and without a timezone (TZ)  will be interpreted as 00:00:00 in GMT.  By default JS dates are formatted to 00:00:00 in the local timezone, this leads to inconsistent results based on how the server is setup.  00:00:00 is also the the behavior of pandas timestamp

Author: "We chose `00:00:00` because that is the design of the pandas Timestamp"

I want the sentence because it shows users our thought process and that we have considered this.

> Date's without an Time of Day (TOD) and without a timezone (TZ)  will be interpreted as 00:00:00 in GMT.  By default JS dates are formatted to 00:00:00 in the local timezone, this leads to inconsistent results based on how the server is setup.  00:00:00 is also the the behavior of pandas timestamp

Reviewer 2: I was just wondering if our design decision to use X internally was relevant here. That's all

Dates with a TOD and without a TZ will be interpreted as that TOD in GMT.  You might want to default to a local timezone here, that option will be added.

Dates with a TOD and TZ will be interpreted with that TOD and TZ.

Dates that can't be explicitly parsed will throw an error and remain as the original string.

Please report bugs where this behavior isn't the case.

## Installing SmartDateField in an existing Faltfile project

> `SmartDateField` is currently in public beta.  This exact code will work currently and against future versions of Flatfile... but we might update the behavior of the `SmartDateField` that is distributed going forward.  This lets us get feedback from users quickly, without locking us into supporting temporary design decisions.

Reviewer 2: ❔ Is this an _alpha_ or _beta_ release?

Author: I'm not sure what we should say here.  This line is really important to get the promise and expectation correct with our users.

Author: Alpha it is

# SmartDateField

## Why

> The `SmartDateField` lets you worry about operations on `Date`s as objects, without worrying about parsing dates, or formatting dates.  Currently the `SmartDateField` can parse [9 date types](./DateField.spec.ts#L43-L60) without any format strings.  This lets you concentrate on business logic, without worrying about incoming date formats.  This let's you confidently perform arithmetic and comparisons in `compute`, `recordCompute`, `batchRecordsCompute`, and `validate` on `Date`s.  `SmartDateField` has an `fString` argument that controls how the date is serialized after `validate` is finished.  Flatfile users no longer have to convert in and out of dates to maintain formatting.

Reviewer 3: Can we change the first "worry about" in the first sentence to "work with"?

The `SmartDateField` lets you work with operations on `Date`s as objects, without worrying about parsing dates, or formatting dates.

Reviewer 4: Agreed with @syshinnn's comment. Also update "This let's you confidently perform arithmetic" to "This lets you confidently perform arithmetic"

Reviewer 4: Also, looks like there are now more date types than just 9 listed here: https://github.com/FlatFilers/platform-sdk-starter/blob/feat/chrono-js-DateField/examples/fields/DateField.spec.ts#L43-L60 

Don't want to undersell ourselves 😉 we should update the number to 16!

Author: done

Dates with a TOD and without a TZ will be interpreted as that TOD in GMT.  You might want to default to a local timezone here, that option will be added.

Dates with a TOD and TZ will be interpreted with that TOD and TZ.

Dates that can't be explicitly parsed will throw an error and remain as the original string.

Please report bugs where this behavior isn't the case.

## Installing SmartDateField in an existing Flatfile project

`SmartDateField` is currently in public beta.  This exact code will work currently and against future versions of Flatfile... but we might update the behavior of the `SmartDateField` that is distributed going forward.  This lets us get feedback from users quickly, without locking us into supporting temporary design decisions.

First copy the whole of [SmartDateField.ts](./SmartDateField.ts) into your repo at `src/SmartDateField.ts`  then in `src/index.ts` you can import `SmartDateField` by adding the line

> `import { SmartDateField } from './SmartDateField'`

Reviewer 3: This might be common sense but I was following these instructions without immediately realizing that it doesn't make sense to put this in index.ts when it should go in my template file.

Author: Done

# SmartDateField

## Why

The `SmartDateField` lets you worry about operations on `Date`s as objects, without worrying about parsing dates, or formatting dates.  Currently the `SmartDateField` can parse [9 date types](./DateField.spec.ts#L43-L60) without any format strings.  This lets you concentrate on business logic, without worrying about incoming date formats.  This let's you confidently perform arithmetic and comparisons in `compute`, `recordCompute`, `batchRecordsCompute`, and `validate` on `Date`s.  `SmartDateField` has an `fString` argument that controls how the date is serialized after `validate` is finished.  Flatfile users no longer have to convert in and out of dates to maintain formatting.

## What

There are comprehensive tests for `SmartDateField`, the most demonstrative being [DateSheet.spec.ts](./DateSheet.spec.ts) .  This shows how dates can be compared as first class objects in a `recordCompute` function, and then formatted to a different persistence format with `fString`.

### Assumptions made

We will add options to override these assumption as needed.

> Date's without an Time of Day (TOD) and without a timezone (TZ)  will be interpreted as 00:00:00 in GMT.  By default JS dates are formatted to 00:00:00 in the local timezone, this leads to inconsistent results based on how the server is setup.  00:00:00 is also the the behavior of pandas timestamp

Reviewer 4: "Dates without a Time of Day (TOD)"

Author: Done

> Dates with a TOD and without a TZ will be interpreted as that TOD in GMT.  You might want to default to a local timezone here, that option will be added.

Reviewer 4: when will the option to default to a local timezone be added? I'd recommend not promising future changes until we are able to release them.

Author: Done

Please report bugs where this behavior isn't the case.

## Installing SmartDateField in an existing Flatfile project

`SmartDateField` is currently in public beta.  This exact code will work currently and against future versions of Flatfile... but we might update the behavior of the `SmartDateField` that is distributed going forward.  This lets us get feedback from users quickly, without locking us into supporting temporary design decisions.

First copy the whole of [SmartDateField.ts](./SmartDateField.ts) into your repo at `src/SmartDateField.ts`  then in `src/index.ts` you can import `SmartDateField` by adding the line

`import { SmartDateField } from './SmartDateField'`

Then:

You must edit the `package.json` at the project root and 
change the line for

 `@flatfile/configure`
to 

`"@flatfile/configure": "^0.4.8",`

additionally you must add the following dependencies:
```
    "dependencies": {
        "chrono-node": "^2.4.1",
        "date-fns": "^2.29.3",
        "date-fns-tz": "^1.3.7"
    }
```

then run `npm install`

Reference the [package.json](../../package.json) include on this branch (soon to be main).

## Using `fString`

> You can control the serialization format of `SmartDateField` with the `fString` option.

Reviewer 5: Remove "serialization" to simplify this sentence.

Reviewer 6: "You can control how dates are converted back to strings with the `fString` option. `fString` represents the date format you want to egress, and specifies how dates should be displayed to the user in the data table."

Author: Done

| DDMonYY      | 09 Feb, 2009      | Day-Month abbreviation-Year with leading zeros      |
| iso full     | Feb142009         | '2009-02-26T00:00:00.000Z'                          |
| D/M/YY       | 22/2/2009         | Day-Month-Year with no leading zeros                |
| DD Mon, YYYY | 16 Feb, 2009      | Day with leading zeros, Month abbreviation, Year    |
| DD/MM/YY     | 18/02/2009        | Day-Month-Year with leading zeros (                 |
| M/D/YY       | 2/21/2009         | Month-Day-Year with no leading zeros                |
| MM/DD/YY     | 02/17/2009        | Month-Day-Year with leading zeros                   |
| Mon DD, YYYY | Feb 15, 2009      | Month abbreviation, Day with leading zeros, Year    |
| MonDDYY      | Feb072009         | Month abbreviation-Day-Year with leading zeros      |
| Month D, Yr  | February 20, 2009 | Month name-Day-Year with no leading zeros           |
| YY/M/D       | 2009/2/23         | Year-Month-Day with no leading zeros                |
| YY/MM/DD     | 2009/02/19        | Year-Month-Day with leading zeros                   |
| bM/bD/YY     | 2/24/2009         | Month-Day-Year with spaces instead of leading zeros |

> 

Author: Colin, I think this is much improved now.  We aren't adding in custom error messaging at this point.

# SmartDateField

## Why

The `SmartDateField` lets you work with operations on `Date`s as objects, without worrying about parsing dates, or formatting dates.  Currently the `SmartDateField` can parse [16 date types](./DateField.spec.ts#L43-L60) without any format strings.  This lets you concentrate on business logic, without worrying about incoming date formats.  This lets you confidently perform arithmetic and comparisons in `compute`, `recordCompute`, `batchRecordsCompute`, and `validate` on `Date`s.  `SmartDateField` has an `fString` argument that controls how the date is serialized after `validate` is finished.  Flatfile users no longer have to convert in and out of dates to maintain formatting.

## What

> There are comprehensive tests for `SmartDateField`, the most demonstrative being [DateSheet.spec.ts](./DateSheet.spec.ts) .  This shows how dates can be compared as first class objects in a `recordCompute` function, and then formatted to a different persistence format with `fString`.

Reviewer 6: ```suggestion
There are comprehensive tests for `SmartDateField`, the most demonstrative being [DateSheet.spec.ts](./DateSheet.spec.ts).  This shows how dates can be compared as first class objects in a `recordCompute` function, and then formatted to a different persistence format with `fString`.
```

### Assumptions made

We will add options to override these assumption as needed.

> Dates without an Time of Day (TOD) and without a timezone (TZ)  will be interpreted as 00:00:00 in GMT.  By default JS dates are formatted to 00:00:00 in the local timezone, this leads to inconsistent results based on how the server is setup.  00:00:00 is also the the behavior of pandas timestamp

Reviewer 6: What's pandas? You haven't introduced that yet. Also missing a period

Dates with a TOD and without a TZ will be interpreted as that TOD in GMT.

Dates with a TOD and TZ will be interpreted with that TOD and TZ.

> Dates that can't be explicitly parsed will throw an error and remain as the original string.

Reviewer 6: Make this a bulleted list

Author: Done

Please report bugs where this behavior isn't the case.

## Installing SmartDateField in an existing Flatfile project

> 

Reviewer 6: ```suggestion

### Note: SmartDateField is in public alpha

```

Author: Done

> `SmartDateField` is currently in public alpha.  This exact code will work currently and against future versions of Flatfile... but we might update the behavior of the `SmartDateField` that is distributed going forward.  This lets us get feedback from users quickly, without locking us into supporting temporary design decisions.

Reviewer 6: ```suggestion
`SmartDateField` is currently in public alpha.  This exact code will work currently and against future versions of Flatfile, but we might update the behavior of the `SmartDateField` that is distributed going forward.  Please send any feedback or ideas for improvement to support@flatfile.com.
```

> First copy the whole of [SmartDateField.ts](./SmartDateField.ts) into your repo at `src/SmartDateField.ts`  then in `src/index.ts` (or relevant template file where you want to use `SmartDateField` you can import `SmartDateField` by adding the line

Reviewer 6: ```suggestion
### How to use

First copy the whole of [SmartDateField.ts](./SmartDateField.ts) into your repo at `src/SmartDateField.ts`  then in `src/index.ts` (or relevant template file where you want to use `SmartDateField` you can import `SmartDateField` by adding the line
```

`import { SmartDateField } from './SmartDateField'`

Then:

You must edit the `package.json` at the project root and 
change the line for

 `@flatfile/configure`
to 

`"@flatfile/configure": "^0.4.8",`

additionally you must add the following dependencies:
```
    "dependencies": {
        "chrono-node": "^2.4.1",
        "date-fns": "^2.29.3",
        "date-fns-tz": "^1.3.7"
    }
```

then run `npm install`

> Reference the [package.json](../../package.json) include on this branch (soon to be main).

Reviewer 6: Remove before merging

Author: Done
